### PR TITLE
Promote Vertex AI FeatureStore resources (GA only)

### PR DESCRIPTION
**Pull request by shotarok**
part of https://github.com/hashicorp/terraform-provider-google/issues/9298

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
`google_vertex_ai_featurestore` (ga only)
`google_vertex_ai_featurestore_entitytype` (ga only)
`google_vertex_ai_featurestore_entitytype_feature` (ga only)
```

### CODE REVIEW

**Comment by shotarok** [comment]
I can't see the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6565/artifacts/c9e9a96c-47f4-40f8-aef3-eb230bac6a32/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-6565/artifacts/c9e9a96c-47f4-40f8-aef3-eb230bac6a32/recording) either due to the permission errors. Is this an expected behavior?

**Comment by shotarok** [comment]
[MEMO] It turns out `monitoringConfig` is only available in beta. I'll put min_version for the field.

**Comment by shotarok** [comment]
Hi @slevenick, could you please review this PR?

**Comment by slevenick** [comment]
Hmmm, it looks like some fields may not be supported in GA at the API level (or were renamed, removed or something else)

This test fails:
TestAccVertexAIFeaturestoreEntitytype_vertexAiFeaturestoreEntitytypeExample

Can you see what is going on with that test? I see this failure message:
```
"description": "Invalid JSON payload received. Unknown name \"monitoringInterval\" at 'entity_type.monitoring_config.snapshot_analysis': Cannot find field.",
```

**Comment by shotarok** [comment]
@slevenick Thank you for your comment! I found out the following things:

* `monitoringInterval` is deprecated in v1beta1 like below
* `stalenessDays` and `monitoringInterval` are supported in both [v1](https://cloud.google.com/vertex-ai/docs/reference/rest/v1/projects.locations.featurestores.entityTypes#FeaturestoreMonitoringConfig) and [v1beta1](https://cloud.google.com/vertex-ai/docs/reference/rest/v1beta1/FeaturestoreMonitoringConfig)

image

I'll remove `monitoringInterval` in this PR and keep this PR only for GA. Apart from that, I'll create another PR to support the new fields. Please let me know if you'd like to contain both changes in this PR. Thanks!

**Comment by shotarok** [comment]
As for the error messages, I'm afraid I couldn't find the visible link to my account. Could you please give me the link to see the error? The following links returned permission errors to me.

> View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6565/artifacts/c9e9a96c-47f4-40f8-aef3-eb230bac6a32/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-6565/artifacts/c9e9a96c-47f4-40f8-aef3-eb230bac6a32/recording) for each test

**Comment by slevenick** [comment]
Yikes, so this is going to be a bit of a tricky situation. We have pretty strict guidelines around backwards compatibility, where we cannot remove a field until a major version change. That means that we can't remove the field from the beta provider, but we also can't have it present in the GA provider. Maybe we can mark that field as min_version: beta to prevent it from being in the GA provider, and then deprecate it so we can remove it in the next major version

**Comment by shotarok** [comment]
> Maybe we can mark that field as min_version: beta to prevent it from being in the GA provider, and then deprecate it so we can remove it in the next major version

Thank you for sharing the release policy. This approach looks good to me. I'll update the PR, and get back to you!

**Comment by shotarok** [comment]
Hello @slevenick, I removed `monitoring_interval` from the example, but `TestAccVertexAIFeaturestoreEntitytype_vertexAiFeaturestoreEntitytypeExample` still is failing. Could you help me what the error message? I'll also try to reproduce the error locally. Thanks

**Review by slevenick — COMMENTED** [review]
Error is:
```
    provider_test.go:315: Step 1/2 error: After applying this test step, the plan was not empty.
        stdout:
        
        
        Terraform used the selected providers to generate the following execution
        plan. Resource actions are indicated with the following symbols:
          ~ update in-place
        
        Terraform will perform the following actions:
        
          # google_vertex_ai_featurestore_entitytype.entity will be updated in-place
          ~ resource "google_vertex_ai_featurestore_entitytype" "entity" {
                id           = "projects/ci-test-project-188019/locations/us-central1/featurestores/terraformxhbvcoal6r/entityTypes/terraformxhbvcoal6r"
                name         = "terraformxhbvcoal6r"
                # (4 unchanged attributes hidden)
        
              + monitoring_config {
                  + snapshot_analysis {
                      + disabled = false
                    }
                }
            }
        
        Plan: 0 to add, 1 to change, 0 to destroy.
--- FAIL: TestAccVertexAIFeaturestoreEntitytype_vertexAiFeaturestoreEntitytypeExample (67.11s)
```

Looks like if monitoring_interval isn't specified we get an empty block back. Can you specify the other field in that test?

**Comment by shotarok** [comment]
Thank you for letting me know about the error! 

The `stalenessDays` and `monitoringIntervalDays` are the other fields, but they're new fields for the provider. The PR becomes not only GA. So I'd try to use `monitoring_config.snapshot_analysis.disabled = true`.

Apart from that, I added `google_vertex_ai_featurestore_entitytype_feature` to the beta provider in https://github.com/GoogleCloudPlatform/magic-modules/pull/6568/. It's also a GA-ready feature store resource. So I'll make it promoted in this PR.

**Comment by slevenick** [comment]
Still seeing:
```
              ~ monitoring_config {
                  ~ snapshot_analysis {
                      - monitoring_interval = "0s" -> null
                        # (1 unchanged attribute hidden)
                    }
                }
            }
```

Are you able to run these tests yourself?

**Comment by shotarok** [comment]
Thank you for your comment. I tried reproducing the error with [the auto-generated branch](https://github.com/modular-magician/terraform-provider-google/commit/b769b21056499b22faf81f5f0bf14a1d74b20042) on my local machine like below, but the test passed on my machine. So I might have overlooked something. I attached [test.log](https://github.com/GoogleCloudPlatform/magic-modules/files/9687987/test.log). 

```
➜  terraform-provider-google git:(auto-pr-6565)  git show HEAD  | head -n 1
commit b769b21056499b22faf81f5f0bf14a1d74b20042

➜  terraform-provider-google git:(auto-pr-6565)  TF_LOG=TRACE make testacc GOOGLE_ORG=test GOOGLE_BILLING_ACCOUNT=test GOOGLE_USE_DEFAULT_CREDENTIALS=true GCLOUD_PROJECT=kouzoh-p-kohama GCLOUD_REGION=us-central1 GCLOUD_ZONE=us-central1-a TEST=./google TESTARGS='-run=TestAccVertexAIFeaturestoreEntitytype_vertexAiFeaturestoreEntitytypeExample' | tee test.log
go generate  ./...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestAccVertexAIFeaturestoreEntitytype_vertexAiFeaturestoreEntitytypeExample -timeout 240m -ldflags="-X=github.com/hashicorp/terraform-provider-google/version.ProviderVersion=acc"
=== RUN   TestAccVertexAIFeaturestoreEntitytype_vertexAiFeaturestoreEntitytypeExample
=== PAUSE TestAccVertexAIFeaturestoreEntitytype_vertexAiFeaturestoreEntitytypeExample
=== CONT  TestAccVertexAIFeaturestoreEntitytype_vertexAiFeaturestoreEntitytypeExample
...
--- PASS: TestAccVertexAIFeaturestoreEntitytype_vertexAiFeaturestoreEntitytypeExample (96.11s)
PASS
ok  	github.com/hashicorp/terraform-provider-google/google	(cached)
```

**Comment by slevenick** [comment]
> Thank you for your comment. I tried reproducing the error with [the auto-generated branch](https://github.com/modular-magician/terraform-provider-google/commit/b769b21056499b22faf81f5f0bf14a1d74b20042) on my local machine like below, but the test passed on my machine. So I might have overlooked something. I attached [test.log](https://github.com/GoogleCloudPlatform/magic-modules/files/9687987/test.log).
> 
> ```
> ➜  terraform-provider-google git:(auto-pr-6565)  git show HEAD  | head -n 1
> commit b769b21056499b22faf81f5f0bf14a1d74b20042
> 
> ➜  terraform-provider-google git:(auto-pr-6565)  TF_LOG=TRACE make testacc GOOGLE_ORG=test GOOGLE_BILLING_ACCOUNT=test GOOGLE_USE_DEFAULT_CREDENTIALS=true GCLOUD_PROJECT=kouzoh-p-kohama GCLOUD_REGION=us-central1 GCLOUD_ZONE=us-central1-a TEST=./google TESTARGS='-run=TestAccVertexAIFeaturestoreEntitytype_vertexAiFeaturestoreEntitytypeExample' | tee test.log
> go generate  ./...
> TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestAccVertexAIFeaturestoreEntitytype_vertexAiFeaturestoreEntitytypeExample -timeout 240m -ldflags="-X=github.com/hashicorp/terraform-provider-google/version.ProviderVersion=acc"
> === RUN   TestAccVertexAIFeaturestoreEntitytype_vertexAiFeaturestoreEntitytypeExample
> === PAUSE TestAccVertexAIFeaturestoreEntitytype_vertexAiFeaturestoreEntitytypeExample
> === CONT  TestAccVertexAIFeaturestoreEntitytype_vertexAiFeaturestoreEntitytypeExample
> ...
> --- PASS: TestAccVertexAIFeaturestoreEntitytype_vertexAiFeaturestoreEntitytypeExample (96.11s)
> PASS
> ok  	github.com/hashicorp/terraform-provider-google/google	(cached)
> ```

It's failing on the beta provider specifically, so maybe it's a beta vs GA difference?

**Comment by shotarok** [comment]
> It's failing on the beta provider specifically, so maybe it's a beta vs GA difference?

Ah, I could reproduce the error with the beta provider locally. I'll look into it. Thanks!

**Comment by shotarok** [comment]
Hi @slevenick I added the default value of the monitoring_interval and I confirmed the tests passed. Could you review the change when you have time? Thanks!

File: mmv1/products/vertexai/api.yaml [hunk 1]
```diff
@@ -273,7 +271,11 @@ objects:
                   The monitoring schedule for snapshot analysis. For EntityType-level config: unset / disabled = true indicates disabled by default for Features under it; otherwise by default enable snapshot analysis monitoring with monitoringInterval for Features under it.
               - !ruby/object:Api::Type::String
                 name: 'monitoringInterval'
+                default_value: "0s"
```
[the remark this answers is not in the document]
Comment (shotarok) [reply mmv1/products/vertexai/api.yaml:274]: Got it. I'll revert the change. BTW, could you give me an example to test a beta field? I'd like to test `monitoring_interval` only in the google-beta.

[the remark this answers is not in the document]
Comment (shotarok) [reply mmv1/products/vertexai/api.yaml:274]: @slevenick Got it! I'll revert the change. BTW, could you give me an example to test a beta field? I'd like to test `monitoring_interval` only in the google-beta.

File: mmv1/products/vertexai/api.yaml [hunk 2]
```diff
@@ -273,7 +271,11 @@ objects:
...
+                min_version: beta
                 description: |
+                  This item is deprecated!
```
[the remark this answers is not in the document]
Comment (shotarok) [reply mmv1/products/vertexai/api.yaml:277]: Thanks, I'll move the message about the deprecation to the field.

File: mmv1/products/vertexai/api.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (shotarok) [reply mmv1/products/vertexai/api.yaml:274]: For example, can we define an additional `google_vertex_ai_featurestore_entitytype` only in the test of the google-beta provider?

```ruby
resource "google_vertex_ai_featurestore_entitytype" "entity" {
  name     = "<%= ctx[:vars]['name'] %>"
  labels = {
    foo = "bar"
  }
  featurestore = google_vertex_ai_featurestore.featurestore.id
  monitoring_config {
    snapshot_analysis {
      disabled = true
    }
  }
}


<%- if is_beta %>
resource "google_vertex_ai_featurestore_entitytype" "entity_for_beta" {
  provider = google-beta
  name     = "<%= ctx[:vars]['name'] + '_for_beta' %>"
  labels = {
    foo = "bar"
  }
  featurestore = google_vertex_ai_featurestore.featurestore.id
  monitoring_config {
    snapshot_analysis {
      disabled = false
      monitoring_interval = "86400s"
    }
  }
}
<%- end %>
```

[the remark this answers is not in the document]
Comment (slevenick) [reply mmv1/products/vertexai/api.yaml:274]: Yes, you can find examples of the version tag needed to make some code only exist in the beta provider. I think its something like <%- unless version == "ga" %>

[the remark this answers is not in the document]
Comment (shotarok) [reply mmv1/products/vertexai/api.yaml:274]: @slevenick it turns out there is no version information in the bindings to render the example. I'm wondering if there is another way to test a beta field. Or should I pass the version to render the example?

https://github.com/GoogleCloudPlatform/magic-modules/blob/3132afe9a47770420119ce2f42cb383ef614b7de/mmv1/provider/terraform/examples.rb#L259-L267

[the remark this answers is not in the document]
Comment (shotarok) [reply mmv1/products/vertexai/api.yaml:274]: I understood we could define multiple examples for the resource, and we can define min_version for an example. I'll add two example erb files.

[the remark this answers is not in the document]
Comment (shotarok) [reply mmv1/products/vertexai/api.yaml:274]: I could define an example only for the beta provider by min_version of an example object. However, I couldn't define an example only for the ga provider only with the min_version. 
https://github.com/shotarok/magic-modules/blob/13f6fa6577c7a72f42ee1a934e6c7ece67e48018/mmv1/templates/terraform/examples/base_configs/test_file.go.erb#L29-L32

File: mmv1/products/vertexai/api.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (shotarok) [reply mmv1/products/vertexai/api.yaml]: I updated the field in [c87fbf6](https://github.com/GoogleCloudPlatform/magic-modules/pull/6565/commits/c87fbf604c11c54c223a15f1c241d1ab3a4e37a6).

File: mmv1/products/vertexai/api.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (shotarok) [reply mmv1/products/vertexai/api.yaml]: I created two examples. One is an example without `monitoring_interval`, which works for both ga and beta. Another is an example with `monitoring_interval`, which only works for beta.

**Comment by shotarok** [comment]
@slevenick OK, the added tests passed in the recording mode. Could you please review this PR again when you have time? Thanks

**Review by slevenick — APPROVED** [review]
Great thanks! Good call on adding a separate test for the beta-only field

**Pull request merged**